# Load Data

In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.width", None)
pd.set_option("display.max_colwidth", None)

transactions = pd.read_csv('../data/h-and-m-personalized-fashion-recommendations/transactions_train.csv')
transactions.head()

,t_dat,customer_id,article_id,price,sales_channel_id
0,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0cad8ffe7ad4a1091e318,663713001,0.050831,2
1,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0cad8ffe7ad4a1091e318,541518023,0.030492,2
2,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699338c5570910a014cc2,505221004,0.015237,2
3,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699338c5570910a014cc2,685687003,0.016932,2
4,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699338c5570910a014cc2,685687004,0.016932,2


In [3]:
customers = pd.read_csv('../data/h-and-m-personalized-fashion-recommendations/customers.csv')
customers.head()

,customer_id,FN,Active,club_member_status,fashion_news_frequency,age,postal_code
0,00000dbacae5abe5e23885899a1fa44253a17956c6d1c3d25f88aa139fdfc657,NaN,NaN,ACTIVE,NONE,49.0,52043ee2162cf5aa7ee79974281641c6f11a68d276429a91f8ca0d4b6efa8100
1,0000423b00ade91418cceaf3b26c6af3dd342b51fd051eec9c12fb36984420fa,NaN,NaN,ACTIVE,NONE,25.0,2973abc54daa8a5f8ccfe9362140c63247c5eee03f1d93f4c830291c32bc3057
2,000058a12d5b43e67d225668fa1f8d618c13dc232df0cad8ffe7ad4a1091e318,NaN,NaN,ACTIVE,NONE,24.0,64f17e6a330a85798e4998f62d0930d14db8db1c054af6c9090f7dd3e38380dc
3,00005ca1c9ed5f5146b52ac8639a40ca9d57aeff4d1bd2c5feb1ca5dff07c43e,NaN,NaN,ACTIVE,NONE,54.0,5d36574f52495e81f019b680c843c443bd343d5ca5b1c222539af5973a23ae6d
4,00006413d8573cd20ed7128e53b7b13819fe5cfc2d801fe7fc0f26dd8d65a85a,1.0,1.0,ACTIVE,Regularly,52.0,25fa5ddee9aac01b35208d01736e57942317d756b32ddd4564743b005a805b1d


In [5]:
articles = pd.read_csv('../data/h-and-m-personalized-fashion-recommendations/articles.csv')
articles.head()

,article_id,product_code,prod_name,product_type_no,product_type_name,product_group_name,graphical_appearance_no,graphical_appearance_name,colour_group_code,colour_group_name,perceived_colour_value_id,perceived_colour_value_name,perceived_colour_master_id,perceived_colour_master_name,department_no,department_name,index_code,index_name,index_group_no,index_group_name,section_no,section_name,garment_group_no,garment_group_name,detail_desc
0,108775015,108775,Strap top,253,Vest top,Garment Upper body,1010016,Solid,9,Black,4,Dark,5,Black,1676,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
1,108775044,108775,Strap top,253,Vest top,Garment Upper body,1010016,Solid,10,White,3,Light,9,White,1676,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
2,108775051,108775,Strap top (1),253,Vest top,Garment Upper body,1010017,Stripe,11,Off White,1,Dusty Light,9,White,1676,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
3,110065001,110065,OP T-shirt (Idro),306,Bra,Underwear,1010016,Solid,9,Black,4,Dark,5,Black,1339,Clean Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Microfibre T-shirt bra with underwired, moulded, lightly padded cups that shape the bust and provide good support. Narrow adjustable shoulder straps and a narrow hook-and-eye fastening at the back. Without visible seams for greater comfort."
4,110065002,110065,OP T-shirt (Idro),306,Bra,Underwear,1010016,Solid,10,White,3,Light,9,White,1339,Clean Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Microfibre T-shirt bra with underwired, moulded, lightly padded cups that shape the bust and provide good support. Narrow adjustable shoulder straps and a narrow hook-and-eye fastening at the back. Without visible seams for greater comfort."


Looks like we won't need 'customers' at this point in our project. I'll leave it behind moving forward.

# Investigate Data

In [6]:
transactions_full = transactions.merge(articles, left_on='article_id', right_on='article_id', how='left')

In [7]:
transactions_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31788324 entries, 0 to 31788323
Data columns (total 29 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   t_dat                         object 
 1   customer_id                   object 
 2   article_id                    int64  
 3   price                         float64
 4   sales_channel_id              int64  
 5   product_code                  int64  
 6   prod_name                     object 
 7   product_type_no               int64  
 8   product_type_name             object 
 9   product_group_name            object 
 10  graphical_appearance_no       int64  
 11  graphical_appearance_name     object 
 12  colour_group_code             int64  
 13  colour_group_name             object 
 14  perceived_colour_value_id     int64  
 15  perceived_colour_value_name   object 
 16  perceived_colour_master_id    int64  
 17  perceived_colour_master_name  object 
 18  department_no       

## Check for any immediate filtering opportunities

In [8]:
# start with all columns
keep = ['t_dat', 'customer_id', 
        #'article_id' <-- used to join, not needed in final df
        'price', 
        # 'sales_channel_id', <-- Out of scope for this project
        'product_code', 'prod_name', 
        'product_type_no', 'product_type_name',
        'product_group_name', 
        'graphical_appearance_no', 'graphical_appearance_name', 
        'colour_group_code', 'colour_group_name',
        'perceived_colour_value_id', 'perceived_colour_value_name',
        'perceived_colour_master_id', 'perceived_colour_master_name',
        'department_no', 'department_name', 
        'index_code', 'index_name',
        'index_group_no', 'index_group_name', 
        'section_no', 'section_name',
        'garment_group_no', 'garment_group_name', 
        'detail_desc']

df_transactions = transactions_full[keep]

In [9]:
df_transactions['t_dat'].max()

'2020-09-22'

In [10]:
df_transactions.head()

,t_dat,customer_id,price,product_code,prod_name,product_type_no,product_type_name,product_group_name,graphical_appearance_no,graphical_appearance_name,colour_group_code,colour_group_name,perceived_colour_value_id,perceived_colour_value_name,perceived_colour_master_id,perceived_colour_master_name,department_no,department_name,index_code,index_name,index_group_no,index_group_name,section_no,section_name,garment_group_no,garment_group_name,detail_desc
0,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0cad8ffe7ad4a1091e318,0.050831,663713,Atlanta Push Body Harlow,283,Underwear body,Underwear,1010016,Solid,9,Black,4,Dark,5,Black,1338,Expressive Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Lace push-up body with underwired, moulded, padded cups for a larger bust and fuller cleavage. Narrow, adjustable shoulder straps, an opening with a hook-and-eye fastening at the back and a lined gusset with concealed press-studs."
1,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0cad8ffe7ad4a1091e318,0.030492,541518,Rae Push (Melbourne) 2p,306,Bra,Underwear,1010016,Solid,51,Light Pink,1,Dusty Light,4,Pink,1334,Casual Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Lace push-up bras with underwired, moulded, padded cups for a larger bust and fuller cleavage. Adjustable shoulder straps and a narrow fastening at the back with two pairs of hooks and eyes."
2,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699338c5570910a014cc2,0.015237,505221,Inca Jumper,252,Sweater,Garment Upper body,1010010,Melange,52,Pink,2,Medium Dusty,4,Pink,5963,Tops Knitwear DS,D,Divided,2,Divided,58,Divided Selected,1003,Knitwear,"Jumper in rib-knit cotton with hard-worn details, a wide, roll-edge neckline, and gently dropped shoulders."
3,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699338c5570910a014cc2,0.016932,685687,W YODA KNIT OL OFFER,252,Sweater,Garment Upper body,1010010,Melange,52,Pink,7,Medium,4,Pink,3090,Campaigns,A,Ladieswear,1,Ladieswear,15,Womens Everyday Collection,1023,Special Offers,"V-neck knitted jumper with long sleeves and ribbing around the neckline, cuffs and hem."
4,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699338c5570910a014cc2,0.016932,685687,W YODA KNIT OL OFFER,252,Sweater,Garment Upper body,1010010,Melange,93,Dark Green,4,Dark,19,Green,3090,Campaigns,A,Ladieswear,1,Ladieswear,15,Womens Everyday Collection,1023,Special Offers,"V-neck knitted jumper with long sleeves and ribbing around the neckline, cuffs and hem."


## Lookup Table Transformation

I think a lot of these columns have a string value, and a numeric value. The numeric value is what we will want to use for our model, so we will need to create a lookup table for the string values.

Below I'll check for unique values in each column, to see if likely pair columns are consistent.

In [11]:
pairs = [
    ('product_code', 'prod_name'),
    ('product_type_no', 'product_type_name'),
    ('graphical_appearance_no', 'graphical_appearance_name'),
    ('colour_group_code', 'colour_group_name'),
    ('perceived_colour_value_id', 'perceived_colour_value_name'),
    ('perceived_colour_master_id', 'perceived_colour_master_name'),
    ('department_no', 'department_name'),
    ('index_code', 'index_name'),
    ('index_group_no', 'index_group_name'),
    ('section_no', 'section_name'),
    ('garment_group_no', 'garment_group_name'),
]

for pair in pairs:
    print(pair[0], pair[1])

    left = np.sort(df_transactions[pair[0]].value_counts().to_numpy())
    right = np.sort(df_transactions[pair[1]].value_counts().to_numpy())
    
    if np.array_equal(left, right):
        print("PAIR")
    else:
        print("NOT PAIR")

product_code prod_name
NOT PAIR
product_type_no product_type_name
NOT PAIR
graphical_appearance_no graphical_appearance_name
PAIR
colour_group_code colour_group_name
PAIR
perceived_colour_value_id perceived_colour_value_name
PAIR
perceived_colour_master_id perceived_colour_master_name
PAIR
department_no department_name
NOT PAIR
index_code index_name
PAIR
index_group_no index_group_name
PAIR
section_no section_name
NOT PAIR
garment_group_no garment_group_name
PAIR


In [12]:
for a, b in pairs:
    vc_a = df_transactions[a].value_counts()
    vc_b = df_transactions[b].value_counts()

    if not np.array_equal(np.sort(vc_a.to_numpy()), np.sort(vc_b.to_numpy())):
        print(f"\nOFF-PAIR: {a} vs {b}")
        print("unique values:", vc_a.shape[0], vc_b.shape[0])
        print("top counts A:", vc_a.head().to_dict())
        print("top counts B:", vc_b.head().to_dict())

        print("count-of-counts A:")
        print(vc_a.value_counts().sort_index().head(10))

        print("count-of-counts B:")
        print(vc_b.value_counts().sort_index().head(10))


OFF-PAIR: product_code vs prod_name
unique values: 46834 45508
top counts A: {706016: 183921, 562245: 158557, 610776: 137799, 599580: 122143, 717490: 81304}
top counts B: {'Jade HW Skinny Denim TRS': 168052, 'Luna skinny RW': 143216, 'Timeless Midrise Brief': 122143, 'Tilly (1)': 105670, 'Cat Tee.': 81304}
count-of-counts A:
count
1     1728
2     1208
3      959
4      854
5      713
6      673
7      568
8      596
9      541
10     431
Name: count, dtype: int64
count-of-counts B:
count
1     1624
2     1131
3      885
4      794
5      680
6      630
7      510
8      562
9      508
10     424
Name: count, dtype: int64

OFF-PAIR: product_type_no vs product_type_name
unique values: 131 130
top counts A: {272: 4217017, 265: 3238428, 252: 2783274, 255: 2203750, 254: 1583408}
top counts B: {'Trousers': 4217017, 'Dress': 3238428, 'Sweater': 2783274, 'T-shirt': 2203750, 'Top': 1583408}
count-of-counts A:
count
4     1
5     1
16    1
20    1
25    1
26    1
28    1
35    2
36    1
41    

In [13]:
off_pairs = [
    ('product_code', 'prod_name'),
    ('product_type_no', 'product_type_name'),
    ('department_no', 'department_name'),
    ('section_no', 'section_name'),
]


def ambiguous_examples(df, left_col, right_col, limit=5):
    left_to_right = (
        df[[left_col, right_col]]
        .drop_duplicates()
        .groupby(left_col)[right_col]
        .agg(lambda s: sorted(s.dropna().astype(str).unique().tolist()))
    )
    right_to_left = (
        df[[left_col, right_col]]
        .drop_duplicates()
        .groupby(right_col)[left_col]
        .agg(lambda s: sorted(s.dropna().astype(str).unique().tolist()))
    )

    left_ambiguous = left_to_right[left_to_right.apply(len) > 1].head(limit)
    right_ambiguous = right_to_left[right_to_left.apply(len) > 1].head(limit)

    return left_ambiguous, right_ambiguous


for left_col, right_col in off_pairs:
    print(f"\n{left_col} vs {right_col}")

    left_ambiguous, right_ambiguous = ambiguous_examples(
        df_transactions, left_col, right_col
    )

    if not left_ambiguous.empty:
        print(f"{left_col} values with multiple {right_col} values:")
        display(left_ambiguous.to_frame(name=f"{right_col}_examples"))

    if not right_ambiguous.empty:
        print(f"{right_col} values with multiple {left_col} values:")
        display(right_ambiguous.to_frame(name=f"{left_col}_examples"))


product_code vs prod_name
product_code values with multiple prod_name values:


,prod_name_examples
product_code,
108775,"[Strap top, Strap top (1)]"
146706,"[3P TANKTOP BODY, Tanktop body white 3PACK]"
156610,"[Anton sport pant, Connor pants]"
160442,"[3p Sneaker Socks, Sneaker 3p Socks]"
179123,"[Long Leggings, Long Leggings (1)]"


prod_name values with multiple product_code values:


,product_code_examples
prod_name,
&DENIM Girlfriend RW,"[618112, 680492]"
1pk embroidery,"[603552, 706271]"
2-PACK SHORTS PRICE,"[596204, 597270]"
2-PACK SKINNY CHEAPO,"[567814, 567821]"
2-p Karlie tee,"[694211, 811278]"



product_type_no vs product_type_name
product_type_name values with multiple product_type_no values:


,product_type_no_examples
product_type_name,
Umbrella,"[532, 83]"



department_no vs department_name
department_name values with multiple department_no values:


,department_no_examples
department_name,
Accessories,"[3510, 3941, 7530, 9985]"
Bags,"[3209, 3944]"
Belts,"[3509, 7389]"
Blazer,"[5131, 5555]"
Blouse,"[1510, 1515, 1522]"



section_no vs section_name
section_name values with multiple section_no values:


,section_no_examples
section_name,
Ladies Other,"[17, 4]"


| Column A        | Column B          | Pair Status | Notes Notes                                                                                                                                                                              | Solution |
| --------------- | ----------------- | ----------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- |
| product_code    | prod_name         | Not Pair    | `product_code` can have multiple `prod_name` spellings, and some `prod_name` values are reused across multiple codes. Keep `product_code` as the lookup key and choose a canonical name. |
| product_type_no | product_type_name | Not Pair    | `Umbrella` appears under two different `product_type_no` values (`83` and `532`), so the name is not unique.                                                                             |
| department_no   | department_name   | Not Pair    | Several `department_name` values map to multiple `department_no` values, including `Jersey`, `Belts`, and `Swimwear`. Use `department_no` as the stable key.                             |
| section_no      | section_name      | Not Pair    | `Ladies Other` maps to two `section_no` values (`4` and `17`), so `section_name` alone is not unique.                                                                                    |


In [14]:
def normalize_to_most_prevalent(df, group_col, value_col):
    pair_counts = (
        df.groupby([group_col, value_col], dropna=False)
        .size()
        .reset_index(name="row_count")
    )
    pair_counts["_tie_break"] = pair_counts[value_col].astype(str)

    canonical = (
        pair_counts.sort_values(
            [group_col, "row_count", "_tie_break"],
            ascending=[True, False, True],
            kind="mergesort",
        )
        .drop_duplicates(subset=[group_col], keep="first")
        [[group_col, value_col, "row_count"]]
        .rename(
            columns={
                value_col: f"canonical_{value_col}",
                "row_count": "canonical_row_count",
            }
        )
    )

    option_counts = (
        pair_counts.groupby(group_col, dropna=False)[value_col]
        .nunique(dropna=False)
        .rename("n_options")
        .reset_index()
    )
    ambiguous_groups = option_counts[option_counts["n_options"] > 1][
        [group_col, "n_options"]
    ]

    updates = ambiguous_groups.merge(canonical, on=group_col, how="left")
    canonical_col = f"canonical_{value_col}"
    update_map = updates.set_index(group_col)[canonical_col]

    original = df[value_col].copy()
    mask = df[group_col].isin(update_map.index)
    df.loc[mask, value_col] = df.loc[mask, group_col].map(update_map)

    changed_rows = (
        ~original.eq(df[value_col]) & ~(original.isna() & df[value_col].isna())
    ).sum()

    return (
        updates.sort_values(["n_options", "canonical_row_count"], ascending=[False, False]),
        int(changed_rows),
    )


# Work on a cleaned copy so the original frame is still available for comparison.
df_transactions_clean = df_transactions.copy()
normalization_log = []

for left_col, right_col in off_pairs:
    for group_col, value_col in [(left_col, right_col), (right_col, left_col)]:
        updates, changed_rows = normalize_to_most_prevalent(
            df_transactions_clean, group_col, value_col
        )
        normalization_log.append(
            {
                "group_col": group_col,
                "value_col": value_col,
                "groups_normalized": len(updates),
                "rows_changed": changed_rows,
            }
        )

        print(f"\n{group_col} -> {value_col}")
        print(f"groups normalized: {len(updates)}")
        print(f"rows changed: {changed_rows}")
        if not updates.empty:
            display(updates.head(10))

normalization_summary = pd.DataFrame(normalization_log)
normalization_summary


product_code -> prod_name
groups normalized: 2165
rows changed: 1827847


,product_code,n_options,canonical_prod_name,canonical_row_count
1391,729928,5,JESSY JERSEY SKIRT,7636
255,501722,5,Arthur slim fit shirt,2199
51,337777,5,Easy iron L/S shirt TP,2196
810,624443,5,ALGOT LS Body 2PACK,973
219,488015,5,TP Stellan shaftless B (1) 7-p,502
601,589521,5,NIKO BOXER YB 5PACK,442
30,293510,5,Basic 2PACK tights YG,375
82,395730,5,MUMS (1),238
811,624543,5,Ronja fancy 5 pack brief,177
206,485726,5,TP Adam 2-p long john BB,116



prod_name -> product_code
groups normalized: 2866
rows changed: 1121361


,prod_name,n_options,canonical_product_code,canonical_row_count
404,Bowie,8,817491,7132
1775,Molly dress,8,879294,5174
2408,Sandy,7,710194,8524
717,Dolly dress,7,778534,8031
1446,Lauren,7,793911,4500
322,Bella dress,7,587191,546
2319,SKINNY LINED,7,738184,62
1719,Melrose,6,675068,16057
2216,Rose,6,685429,11398
2227,Ruby,6,876053,8885



product_type_no -> product_type_name
groups normalized: 0
rows changed: 0

product_type_name -> product_type_no
groups normalized: 1
rows changed: 2111


,product_type_name,n_options,canonical_product_type_no,canonical_row_count
0,Umbrella,2,532,2939



department_no -> department_name
groups normalized: 0
rows changed: 0

department_name -> department_no
groups normalized: 22
rows changed: 4344790


,department_name,n_options,canonical_department_no,canonical_row_count
11,Knitwear,11,1626,856179
20,Trouser,7,1722,1161125
7,Jersey,5,1660,465274
16,Shorts,5,1723,329635
15,Shirt,4,5777,53504
0,Accessories,4,9985,36129
4,Blouse,3,1522,908849
5,Dress,3,1322,768025
9,Jersey Fancy,3,5828,94402
19,Swimwear,2,4242,2455138



section_no -> section_name
groups normalized: 0
rows changed: 0

section_name -> section_no
groups normalized: 1
rows changed: 434


,section_name,n_options,canonical_section_no,canonical_row_count
0,Ladies Other,2,4,583


,group_col,value_col,groups_normalized,rows_changed
0,product_code,prod_name,2165,1827847
1,prod_name,product_code,2866,1121361
2,product_type_no,product_type_name,0,0
3,product_type_name,product_type_no,1,2111
4,department_no,department_name,0,0
5,department_name,department_no,22,4344790
6,section_no,section_name,0,0
7,section_name,section_no,1,434


In [15]:
for pair in pairs:
    print(pair[0], pair[1])

    left = np.sort(df_transactions_clean[pair[0]].value_counts().to_numpy())
    right = np.sort(df_transactions_clean[pair[1]].value_counts().to_numpy())
    
    if np.array_equal(left, right):
        print("PAIR")
    else:
        print("NOT PAIR")

product_code prod_name
PAIR
product_type_no product_type_name
PAIR
graphical_appearance_no graphical_appearance_name
PAIR
colour_group_code colour_group_name
PAIR
perceived_colour_value_id perceived_colour_value_name
PAIR
perceived_colour_master_id perceived_colour_master_name
PAIR
department_no department_name
PAIR
index_code index_name
PAIR
index_group_no index_group_name
PAIR
section_no section_name
PAIR
garment_group_no garment_group_name
PAIR


Great. Now we're ready to split into a lookup table.

In [16]:
pairs = [# 'product_code', 'prod_name', <-- too granular for this scope
    'product_type_no', 'product_type_name',
    'graphical_appearance_no', 'graphical_appearance_name',
    'colour_group_code', 'colour_group_name',
    'perceived_colour_value_id', 'perceived_colour_value_name',
    'perceived_colour_master_id', 'perceived_colour_master_name',
    # 'department_no', 'department_name', <-- h&m exclusive
    # 'index_code', 'index_name', <-- out of scope; too granular
    'index_group_no', 'index_group_name',
    # 'section_no', 'section_name', <-- not going to be helpful, h&m exlusive
    'garment_group_no', 'garment_group_name'
]

df_lookup = df_transactions_clean[pairs].drop_duplicates()

In [ ]:
# lookup_cols = pairs + 
want = [
    't_dat', # date of transaction
    'price', # price of item
    'product_type_no', # product type (Specific, ID)
    'product_group_name', # product group (General) 
    'graphical_appearance_no', 'graphical_appearance_name', # graphical appearance
    'colour_group_code', # color group
    'perceived_colour_value_id', # perceived color value
    'perceived_colour_master_id', # master color
    # 'department_no', 'department_name', 
    # 'index_code', 'index_name',
    'index_group_no', 'index_group_name', # Gender
    # 'section_no', 'section_name',
    'garment_group_no', 'garment_group_name', 
    # 'detail_desc'
]

keep_mapping = {
    't_dat': 'date',
    'product_type_no': 'product_type_id',
    'product_group_name': 'product_group_id',
    'graphical_appearance_no': 'graphical_appearance_id',
    'perceived_colour_value_id': 'color_specific_id',
    'perceived_colour_master_id': 'color_master_id',
    'index_group_no': 'gender_id'
}

lookup_keep = [
    'product_type_id', 'product_type_name', 
    'product_group_id', 'product_group_name',
    'graphical_appearance_id', 'graphical_appearance_name', 
    'color_specific_id', 'color_specific_name',
    'color_master_id', 'color_master_name',
    'gender_id', 'index_group_name', 
]

lookup_keep_mapping = {
    'product_type_no': 'product_type_id',
    'product_group_name': 'product_group_id',
    'graphical_appearance_no': 'graphical_appearance_id',
    'perceived_colour_value_id': 'color_specific_id',
    'perceived_colour_master_id': 'color_master_id',
    'index_group_no': 'gender_id'
}

In [ ]:
df_transactions = df_transactions.rename(columns=keep_mapping)

In [24]:
df_transactions.head()

,t_dat,customer_id,price,product_code,prod_name,product_type_no,product_type_name,product_group_name,graphical_appearance_no,graphical_appearance_name,colour_group_code,colour_group_name,perceived_colour_value_id,perceived_colour_value_name,perceived_colour_master_id,perceived_colour_master_name,department_no,department_name,index_code,index_name,index_group_no,index_group_name,section_no,section_name,garment_group_no,garment_group_name,detail_desc
0,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0cad8ffe7ad4a1091e318,0.050831,663713,Atlanta Push Body Harlow,283,Underwear body,Underwear,1010016,Solid,9,Black,4,Dark,5,Black,1338,Expressive Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Lace push-up body with underwired, moulded, padded cups for a larger bust and fuller cleavage. Narrow, adjustable shoulder straps, an opening with a hook-and-eye fastening at the back and a lined gusset with concealed press-studs."
1,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0cad8ffe7ad4a1091e318,0.030492,541518,Rae Push (Melbourne) 2p,306,Bra,Underwear,1010016,Solid,51,Light Pink,1,Dusty Light,4,Pink,1334,Casual Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Lace push-up bras with underwired, moulded, padded cups for a larger bust and fuller cleavage. Adjustable shoulder straps and a narrow fastening at the back with two pairs of hooks and eyes."
2,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699338c5570910a014cc2,0.015237,505221,Inca Jumper,252,Sweater,Garment Upper body,1010010,Melange,52,Pink,2,Medium Dusty,4,Pink,5963,Tops Knitwear DS,D,Divided,2,Divided,58,Divided Selected,1003,Knitwear,"Jumper in rib-knit cotton with hard-worn details, a wide, roll-edge neckline, and gently dropped shoulders."
3,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699338c5570910a014cc2,0.016932,685687,W YODA KNIT OL OFFER,252,Sweater,Garment Upper body,1010010,Melange,52,Pink,7,Medium,4,Pink,3090,Campaigns,A,Ladieswear,1,Ladieswear,15,Womens Everyday Collection,1023,Special Offers,"V-neck knitted jumper with long sleeves and ribbing around the neckline, cuffs and hem."
4,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699338c5570910a014cc2,0.016932,685687,W YODA KNIT OL OFFER,252,Sweater,Garment Upper body,1010010,Melange,93,Dark Green,4,Dark,19,Green,3090,Campaigns,A,Ladieswear,1,Ladieswear,15,Womens Everyday Collection,1023,Special Offers,"V-neck knitted jumper with long sleeves and ribbing around the neckline, cuffs and hem."


In [ ]:
df_transactions.drop(columns=['customer_id', ])

In [43]:
df_transactions['garment_group_name'].value_counts()

garment_group_name
Jersey Fancy                     5171611
Jersey Basic                     3268451
Under-, Nightwear                2966788
Trousers                         2953711
Swimwear                         2538085
Blouses                          2510441
Knitwear                         2392807
Dresses Ladies                   2147529
Accessories                      1645464
Trousers Denim                   1279886
Socks and Tights                  747344
Shorts                            736716
Shoes                             729477
Outdoor                           715576
Skirts                            624822
Dressed                           409535
Unknown                           391639
Special Offers                    232397
Shirts                            210289
Woven/Jersey/Knitted mix Baby      59863
Dresses/Skirts girls               55893
Name: count, dtype: int64

In [34]:
df_transactions['product_group_name'].value_counts()

product_group_name
Garment Upper body       12552755
Garment Lower body        7046054
Garment Full body         3552470
Swimwear                  2579222
Underwear                 2565858
Accessories               1599593
Shoes                      745521
Socks & Tights             685712
Nightwear                  348180
Unknown                     97040
Bags                         7313
Items                        5427
Cosmetic                     1500
Underwear/nightwear           559
Furniture                     533
Garment and Shoe care         279
Stationery                    229
Interior textile               74
Fun                             5
Name: count, dtype: int64

### Create lookup column for 'product_group_name'

...

### Cut baby products

...